**So first, I imported pandas because it's my go-to tool for reading and handling CSV files. I then read the 2021 dataset using pd.read_csv() and just checked out the first few rows with df.head() to get a feel for what the data looks like. I also run the code df.isnull().sum() and df.dtypes to see if there were any missing values or data types I should know about.**

In [2]:
import pandas as pd
file_path = "PRPL-Facebook-Government-Report-2021_H1.csv"
df = pd.read_csv(file_path)

In [3]:
df.head()

Country Preservation Requests Preservation Accounts Preserved  \
0     Afghanistan                     0                               0   
1   Aland Islands                     0                               0   
2         Albania                    81                             117   
3         Algeria                     0                               0   
4  American Samoa                     0                               0   

  Total Requests Number of Requests Where Some Data Produced  \
0              0                                           0   
1              0                                           0   
2             24                                          17   
3              0                                           0   
4              0                                           0   

  Total Requests Accounts Total Requests Percentage  \
0                       0                        0%   
1                       0                        0%   
2                      63                       71%   
3                       0                        0%   
4                       0                        0%   

  Legal Process Request Total  \
0                           0   
1                           0   
2                          24   
3                           0   
4                           0   

  Legal Number of Requests Where Some Data Produced  \
0                                                 0   
1                                                 0   
2                                                17   
3                                                 0   
4                                                 0   

  Legal Process Request Total Accounts  ... Title III Percentage  \
0                                    0  ...                  NaN   
1                                    0  ...                  NaN   
2                                   63  ...                  NaN   
3                                    0  ...                  NaN   
4                                    0  ...                  NaN   

  FISA Content Requests FISA Content Requests Accounts  \
0                   NaN                            NaN   
1                   NaN                            NaN   
2                   NaN                            NaN   
3                   NaN                            NaN   
4                   NaN                            NaN   

  FISA Content Requests Percentage FISA Non-Content Requests  \
0                              NaN                       NaN   
1                              NaN                       NaN   
2                              NaN                       NaN   
3                              NaN                       NaN   
4                              NaN                       NaN   

  FISA Non-Content Requests Accounts FISA Non-Content Requests Percentage  \
0                                NaN                                  NaN   
1                                NaN                                  NaN   
2                                NaN                                  NaN   
3                                NaN                                  NaN   
4                                NaN                                  NaN   

  NSLs NSLs Accounts NSLs Percentage  
0  NaN           NaN             NaN  
1  NaN           NaN             NaN  
2  NaN           NaN             NaN  
3  NaN           NaN             NaN  
4  NaN           NaN             NaN  

[5 rows x 48 columns]

In [4]:
df.isnull().sum()

Country                                                0
Preservation Requests                                  0
Preservation Accounts Preserved                        0
Total Requests                                         0
Number of Requests Where Some Data Produced            0
Total Requests Accounts                                0
Total Requests Percentage                              0
Legal Process Request Total                            0
Legal Number of Requests Where Some Data Produced      0
Legal Process Request Total Accounts                   0
Legal Process Request Total Percentage                 0
Emergency Request Total                                0
ER Number of Requests Where Some Data Produced         0
Emergency Request Total Accounts                       0
Emergency Request Total Percentage                     0
Search Warrant                                       250
Search Warrant Accounts                              250
Search Warrant Percentage      

In [5]:
df.dtypes

Country                                               object
Preservation Requests                                 object
Preservation Accounts Preserved                       object
Total Requests                                        object
Number of Requests Where Some Data Produced           object
Total Requests Accounts                               object
Total Requests Percentage                             object
Legal Process Request Total                           object
Legal Number of Requests Where Some Data Produced     object
Legal Process Request Total Accounts                  object
Legal Process Request Total Percentage                object
Emergency Request Total                               object
ER Number of Requests Where Some Data Produced        object
Emergency Request Total Accounts                      object
Emergency Request Total Percentage                    object
Search Warrant                                        object
Search Warrant Accounts 

**Here’s where I started the cleaning process.I noticed that from the column 'Search Warrant' to 'NSLs Percentage', I didn’t really need that range of data for my analysis. So, instead of listing every column manually, I used get_loc() to find the start and end column indexes and dropped everything in between. This helped me keep only the useful parts of the dataset.**

In [6]:
file_path = "PRPL-Facebook-Government-Report-2021_H1.csv"
df = pd.read_csv(file_path)

start_col = 'Search Warrant'
end_col = 'NSLs Percentage'
start_index = df.columns.get_loc(start_col)
end_index = df.columns.get_loc(end_col)
columns_to_drop = df.columns[start_index:end_index+1].tolist()
df = df.drop(columns=columns_to_drop)

output_file_path = "PRPL-Facebook-Government-Report-2021_H1_modified.csv"
df.to_csv(output_file_path, index=False)

**After dropping those columns,I saved the cleaned version of the 2021 dataset into a new CSV file named "PRPL-Facebook-Government-Report-2021_H1_modified.csv".This way,I have a separate cleaned version that I can reuse without running the code again every time.**

In [7]:
file_path = "PRPL-Facebook-Government-Report-2021_H1_modified.csv"
df = pd.read_csv(file_path)
df.columns

Index(['Country', 'Preservation Requests', 'Preservation Accounts Preserved',
       'Total Requests', 'Number of Requests Where Some Data Produced',
       'Total Requests Accounts', 'Total Requests Percentage',
       'Legal Process Request Total',
       'Legal Number of Requests Where Some Data Produced',
       'Legal Process Request Total Accounts',
       'Legal Process Request Total Percentage', 'Emergency Request Total',
       'ER Number of Requests Where Some Data Produced',
       'Emergency Request Total Accounts',
       'Emergency Request Total Percentage'],
      dtype='object')

In [8]:
df.isnull().sum()

Country                                              0
Preservation Requests                                0
Preservation Accounts Preserved                      0
Total Requests                                       0
Number of Requests Where Some Data Produced          0
Total Requests Accounts                              0
Total Requests Percentage                            0
Legal Process Request Total                          0
Legal Number of Requests Where Some Data Produced    0
Legal Process Request Total Accounts                 0
Legal Process Request Total Percentage               0
Emergency Request Total                              0
ER Number of Requests Where Some Data Produced       0
Emergency Request Total Accounts                     0
Emergency Request Total Percentage                   0
dtype: int64

**Here,I followed the same process for 2020, 2022, 2023, and 2024.For each year,I cleaned the data the same way (dropping the unnecessary columns and saving the cleaned version to a new file).This gave me five cleaned datasets ready to be merged. Now that I had all five cleaned CSVs,I wanted to read each one back into Python using pd.read_csv() and stored them in variables like df_1, df_2, etc. This set me up to combine them all into one big dataset.**

In [9]:
df_1 = pd.read_csv("PRPL-Facebook-Government-Report-2020_H1_modified.csv")
df_2 = pd.read_csv("PRPL-Facebook-Government-Report-2021_H1_modified.csv")
df_3 = pd.read_csv("PRPL-Facebook-Government-Report-2022_H1_modified.csv")
df_4 = pd.read_csv("PRPL-Facebook-Government-Report-2023_H1_modified.csv")
df_5 = pd.read_csv("PRPL-Facebook-Government-Report-2024_H1_modified.csv")

In [10]:
combined_df = pd.concat([df_1, df_2, df_3, df_4, df_5], ignore_index=True)
combined_df.head(100)

Country Preservation Requests Preservation Accounts Preserved  \
0      Afghanistan                     0                               0   
1    Aland Islands                     0                               0   
2          Albania                     9                               9   
3          Algeria                     1                               1   
4   American Samoa                     0                               0   
..             ...                   ...                             ...   
95        Honduras                    21                              48   
96       Hong Kong                    47                              50   
97         Hungary                    24                              26   
98         Iceland                    13                              15   
99           India                 4,132                           9,406   

   Total Requests Number of Requests Where Some Data Produced  \
0               0                                           0   
1               0                                           0   
2              11                                           4   
3               2                                           0   
4               0                                           0   
..            ...                                         ...   
95              9                                           4   
96            262                                          62   
97            409                                         203   
98              1                                           1   
99         35,560                                      17,922   

   Total Requests Accounts Total Requests Percentage  \
0                        0                        0%   
1                        0                        0%   
2                       24                       36%   
3                        3                        0%   
4                        0                        0%   
..                     ...                       ...   
95                      23                       44%   
96                     285                       24%   
97                     593                       50%   
98                       1                      100%   
99                  57,294                       50%   

   Legal Process Request Total  \
0                            0   
1                            0   
2                           10   
3                            1   
4                            0   
..                         ...   
95                           3   
96                         259   
97                         406   
98                           1   
99                      33,374   

   Legal Number of Requests Where Some Data Produced  \
0                                                  0   
1                                                  0   
2                                                  3   
3                                                  0   
4                                                  0   
..                                               ...   
95                                                 1   
96                                                61   
97                                               203   
98                                                 1   
99                                            16,781   

   Legal Process Request Total Accounts  \
0                                     0   
1                                     0   
2                                    23   
3                                     2   
4                                     0   
..                                  ...   
95                                   13   
96                                  282   
97                                  587   
98                                    1   
99                               49,290   

   Legal Process Request Total Percentage Eme

**Before merging them,I wanted each row to show which year it came from. I wrote a small function called extract_year() to pull the year from the filename (since the year is already part of the file name).Then I applied that function to each DataFrame, creating a new 'Year' column in all five.**

In [11]:
def extract_year(filename): 
    year_part = filename.split('-')[4]
    return int(year_part[:4])

df_1['Year'] = extract_year("PRPL-Facebook-Government-Report-2020_H1_modified.csv")
df_2['Year'] = extract_year("PRPL-Facebook-Government-Report-2021_H1_modified.csv")
df_3['Year'] = extract_year("PRPL-Facebook-Government-Report-2022_H1_modified.csv")
df_4['Year'] = extract_year("PRPL-Facebook-Government-Report-2023_H1_modified.csv")
df_5['Year'] = extract_year("PRPL-Facebook-Government-Report-2024_H1_modified.csv")


**Once every file had the 'Year' column, I used pd.concat() to combine all five cleaned DataFrames into a single combined_df. I used ignore_index=True to make sure the index values reset properly across the merged dataset.**

In [12]:
combined_df = pd.concat([df_1, df_2, df_3, df_4, df_5], ignore_index=True)
combined_df.to_csv("combined_file_with_year.csv", index=False)

**Finally, I saved the fully combined DataFrame into a new CSV called "Combined_file_with_year.csv".This file contains all the cleaned and labeled data from 2020 to 2024, ready for analysis, visualization. I loaded the merged file and checked the first 300 rows using df.head(300),just to confirm everything looked perfect and all years were properly included.**

In [15]:
file_path = "combined_file_with_year.csv"
df = pd.read_csv(file_path)
df.head(300)

Country Preservation Requests  \
0                Afghanistan                     0   
1              Aland Islands                     0   
2                    Albania                     9   
3                    Algeria                     1   
4             American Samoa                     0   
..                       ...                   ...   
295                    Chile                   416   
296                    China                     0   
297         Christmas Island                     0   
298  Cocos (Keeling) Islands                     0   
299                 Colombia                   958   

    Preservation Accounts Preserved Total Requests  \
0                                 0              0   
1                                 0              0   
2                                 9             11   
3                                 1              2   
4                                 0              0   
..                              ...            ...   
295                             865            348   
296                               0              0   
297                               0              0   
298                               0              0   
299                           1,596            722   

    Number of Requests Where Some Data Produced Total Requests Accounts  \
0                                             0                       0   
1                                             0                       0   
2                                             4                      24   
3                                             0                       3   
4                                             0                       0   
..                                          ...                     ...   
295                                          52                     756   
296                                           0                       0   
297                                           0                       0   
298                                           0                       0   
299                                         426                   1,702   

    Total Requests Percentage Legal Process Request Total  \
0                          0%                           0   
1                          0%                           0   
2                         36%                          10   
3                          0%                           1   
4                          0%                           0   
..                        ...                         ...   
295                       15%                         295   
296                        0%                           0   
297                        0%                           0   
298                        0%                           0   
299                       59%                         421   

    Legal Number of Requests Where Some Data Produced  \
0                                                   0   
1                                                   0   
2                                                   3   
3                                                   0   
4                                                   0   
..                                                ...   
295                                                19   
296                                                 0   
297                                                 0   
298                                                 0   
299                                               234   

    Legal Process Request Total Accounts  \
0                                      0   
1                                      0   
2                                     23   
3                                      2   
4                                      0   
..                                   ...   
295                                  626   
296                                    0   
297     

**I imported matplotlib.pyplot and seaborn for some basic plotting if needed, and plotly.express for creating those cool interactive visualizations.And checked the data types and column names just to make sure everything looked good and ready for transformation.**

In [14]:
from matplotlib import pyplot as plt 
import seaborn as sns
import plotly.express as px

In [16]:
df.dtypes

Country                                              object
Preservation Requests                                object
Preservation Accounts Preserved                      object
Total Requests                                       object
Number of Requests Where Some Data Produced          object
Total Requests Accounts                              object
Total Requests Percentage                            object
Legal Process Request Total                          object
Legal Number of Requests Where Some Data Produced    object
Legal Process Request Total Accounts                 object
Legal Process Request Total Percentage               object
Emergency Request Total                              object
ER Number of Requests Where Some Data Produced       object
Emergency Request Total Accounts                     object
Emergency Request Total Percentage                   object
Year                                                  int64
dtype: object

In [18]:
df.columns

Index(['Country', 'Preservation Requests', 'Preservation Accounts Preserved',
       'Total Requests', 'Number of Requests Where Some Data Produced',
       'Total Requests Accounts', 'Total Requests Percentage',
       'Legal Process Request Total',
       'Legal Number of Requests Where Some Data Produced',
       'Legal Process Request Total Accounts',
       'Legal Process Request Total Percentage', 'Emergency Request Total',
       'ER Number of Requests Where Some Data Produced',
       'Emergency Request Total Accounts',
       'Emergency Request Total Percentage', 'Year'],
      dtype='object')

**Now, this part is where I wanted to make sure all the numerical columns were actually in a clean, usable format. Some of them had commas (like "1,234"), which can mess up calculations, so I created a list of all the number-related columns.I ran a loop where I first converted each column to a string (to deal with the commas),removed those commas using str.replace(),converted them to floats (just in case there were decimal points),and then finally to integers.I also filled any missing values with 0 during this process.Doing this just made the data more reliable for analysis and visualization.**

**Similarly, I made a separate list for percentage columns. These had % signs in them, so again, I looped through and turned each into a string,stripped out the % symbols, and then converted everything into floats. This way,I could use them easily in calculations or charts later on if needed.Keeping this separate from the number columns made the code cleaner and easier to follow.**

In [17]:
number_columns = [
    "Preservation Requests",
    "Preservation Accounts Preserved",
    "Total Requests",
    "Number of Requests Where Some Data Produced",
    "Total Requests Accounts",
    "Legal Process Request Total",
    "Legal Number of Requests Where Some Data Produced",
    "Legal Process Request Total Accounts",
    "Emergency Request Total",
    "ER Number of Requests Where Some Data Produced",
    "Emergency Request Total Accounts"
]

for col in number_columns:
    df[col] = df[col].astype(str)              
    df[col] = df[col].str.replace(',', '')     
    df[col] = df[col].astype(float)             
    df[col] = df[col].fillna(0).astype(int)    

percent_columns = [
    "Total Requests Percentage",
    "Legal Process Request Total Percentage",
    "Emergency Request Total Percentage"
]

for col in percent_columns:
    df[col] = df[col].astype(str)              
    df[col] = df[col].str.replace('%', '')     
    df[col] = df[col].astype(float)       

df.dtypes

Country                                               object
Preservation Requests                                  int64
Preservation Accounts Preserved                        int64
Total Requests                                         int64
Number of Requests Where Some Data Produced            int64
Total Requests Accounts                                int64
Total Requests Percentage                            float64
Legal Process Request Total                            int64
Legal Number of Requests Where Some Data Produced      int64
Legal Process Request Total Accounts                   int64
Legal Process Request Total Percentage               float64
Emergency Request Total                                int64
ER Number of Requests Where Some Data Produced         int64
Emergency Request Total Accounts                       int64
Emergency Request Total Percentage                   float64
Year                                                   int64
dtype: object

**Once the data was all cleaned up, I moved on to visualizing! First, I made a bar chart using Plotly to show how "Total Requests" changed over the years.Then,I grouped the data by year and summed up the total requests,just to get a cleaner, year-by-year breakdown. After that, I recreated the bar chart, this time using the grouped data. It gave me a more focused view of how the total number of requests evolved across the five years.**

In [ ]:
yearly_requests = df.groupby("Year")["Total Requests"].sum().reset_index()

fig_bar = px.bar(
    yearly_requests,
    x="Year",
    y="Total Requests",
    text="Total Requests",
    color="Total Requests",
    color_continuous_scale="Blues",
    title=" Total Requests Over the Years"
)

fig_bar.show()

In [73]:
import plotly.graph_objects as go

**I created a line plot to show how different types of requests (Legal Process, Emergency, and Preservation) changed over time.I started by grouping the data by year and calculated the sum of the three major request types.I visualized them using a line chart with markers so I could clearly see how each type of request evolved over time. Then I customized the layout,added year labels on the x-axis, a nice light grey background, bigger font for readability, and gave the legend a proper title. My goal was to help the audience quickly spot patterns,like which request types are increasing faster or where sudden changes happened.**

In [ ]:
request_types = df.groupby("Year")[
    ["Legal Process Request Total", "Emergency Request Total", "Preservation Requests"]
].sum().reset_index()

fig_line = px.line(
    request_types,
    x="Year",
    y=["Legal Process Request Total", "Emergency Request Total", "Preservation Requests"],
    markers=True,
    title="Request Types Over the Years"
)

fig_line.update_layout(
    xaxis=dict(
        tickmode='array',
        tickvals=request_types['Year'],
        ticktext=[str(year) for year in request_types['Year']],
        title="Year",
        showgrid=True,           
        gridcolor='white'      
    ),
    yaxis=dict(
        title="Total Requests",
        showgrid=True,            
        gridcolor='white'     
    ),
    plot_bgcolor="lightgrey",
    title_x=0.5,
    font=dict(size=14),
    legend_title_text="Request Type"
)

fig_line.show()

**Next, I wanted to understand the global distribution of the requests, so I grouped the total requests by country. I created a choropleth map using Plotly’s px.choropleth, which basically points each country a different shade of blue depending on how many total requests they had.It gave me a global view,really helpful for spotting which countries are getting more attention in these reports.I chose a blue color scale where darker shades represent more requests. Here, I wanted give the audience a quick, intuitive feel for the global distribution of requests highlighting the countries with the heaviest activity.For color part,I wanted to chose blue because it's professional, neutral, and doesn’t trigger strong emotions. Red or green might seem alarming or too positive. Blue is clean and helps users focus on the data instead of getting distracted by colors.And don't want to use all the color in graph, because it makes graph look messy.**

In [ ]:
df['Total Requests'] = pd.to_numeric(df['Total Requests'])
country_requests = df.groupby('Country')["Total Requests"].sum().reset_index()

fig_map = px.choropleth(
    country_requests,
    locations="Country",
    locationmode="country names",
    color="Total Requests",
    color_continuous_scale="Blues",
    title="Total Requests by Country (All Years Combined)",
)

fig_map.update_layout(
    geo=dict(showframe=False, showcoastlines=True)
)

fig_map.show()

**After that,I thought,“Wouldn’t it be cool to see how this changed over time?” So I made an animated choropleth map where each frame represents a different year.This way, I could watch how total requests by country changed year by year.It made patterns more dynamic and easier to notice, like how some countries might’ve seen spikes in certain years.**

In [ ]:
df['Total Requests'] = pd.to_numeric(df['Total Requests'])

country_year_requests = df.groupby(['Year', 'Country'])['Total Requests'].sum().reset_index()

fig_map_y = px.choropleth(
    country_year_requests,
    locations="Country",
    locationmode="country names",
    color="Total Requests",
    animation_frame="Year",
    color_continuous_scale="Blues",
    title="Total Requests by Country Over the Years",
)

fig_map_y.update_layout(
    geo=dict(showframe=False, showcoastlines=True),
    title_x=0.5
)

fig_map_y.show()


**After showing total request by country over the year, I wanted to dig a little deeper, Which type of request is more dominant in each country? To do this, I grouped the data by country and request type (Preservation, Legal Process, Emergency).I then created a choropleth map using Plotly’s px.choropleth, which shades each country based on the number of total requests. This map provided a global view, making it easier to spot which countries had the most activity. I chose a blue color scale because it’s neutral, and visually calming. Blue helps users focus on the data without being distracted by overly bold colors like red or green. Additionally, I kept the color scale simple to avoid cluttering the map, making it clean and easy to interpret. I used hover becuase it is simple way to display additional information when users what to interact with the map. Here in the map by using hover, they can instantly see the total number of requests for that country without cluttering the map with text or labels.I want to make the audience explore how different request types are distributed globally, and easily spot patterns depending on the request type.** 

In [ ]:
df['Total Requests'] = df[["Preservation Requests", "Legal Process Request Total", "Emergency Request Total"]].sum(axis=1)
country_requests = df.groupby('Country')[
    ["Preservation Requests", "Legal Process Request Total", "Emergency Request Total", "Total Requests"]
].sum().reset_index()

fig_rt = px.choropleth(
    country_requests,
    locations="Country",
    locationmode="country names",
    color="Total Requests", 
    color_continuous_scale="Blues",
    title="Requests by Country and Type",
    hover_name="Country", 
    hover_data=["Total Requests"] 
)

fig_rt.update_layout(
    updatemenus=[
        dict(buttons=[dict(
                    label=col,
                    method="restyle",
                    args=[{"z": [country_requests[col]]}, 
                          {"title": f"Requests by Country: {col}",
                           "hover_data": [col, "Total Requests"]}] 
                )
                for col in ["Preservation Requests", "Legal Process Request Total", "Emergency Request Total"]
            ],
            direction="down",
            showactive=True,
            x=0.1,
            y=1.1
        )
    ],
    title_x=0.5,
    geo=dict(showframe=False, showcoastlines=True)
)

fig_rt.show()


**At this stage, I was interested in expanding the image and provide a closer look at a few key countries. I chose ten countries, including the US, India, Germany, Japan, and others, rather than displaying every nation, which could be overwhelming. I decided on a heatmap since it's great for quickly identifying patterns.After that,I used a pivot_table() to organize the data into a grid with countries as rows, years as columns, and total requests as the values. This made it easy to plot in the heatmap format, where I could see at a glance which countries and years had higher or lower requests based on the color scale.This visual makes it really easy for the audience to immediately see which countries had spikes in requests, which stayed steady, and which ones suddenly dropped off. Let the audience quickly compare countries side by side over the years, and spot interesting patterns like sudden surges, consistent growth, or unexpected drops. For Color code 'RdYlBu_r' I took help of google.**

In [ ]:
selected_countries = ['Australia', 'Brazil', 'Canada', 'France', 'Germany', 
                      'India', 'Japan', 'South Korea', 'United Kingdom', 'United States']

df_selected = df[df['Country'].isin(selected_countries)]

heatmap_data = df_selected.pivot_table(
    index='Country',
    columns='Year',
    values='Total Requests',
    aggfunc='sum'
)

fig_heatmap = px.imshow(
    heatmap_data,
    labels=dict(x="Year", y="Country", color="Total Requests"),
    color_continuous_scale='RdYlBu_r', 
    title='Heatmap of Total Requests by Country and Year'
)

fig_heatmap.update_layout(
    title_x=0.5,
    xaxis_title="Year",
    yaxis_title="Country"
)

fig_heatmap.show()


**I realized the heatmap could be more readable if the countries were sorted from lowest to highest total requests. So I grouped the data again, summed up total requests per country, sorted them in ascending order, and rearranged the heatmap rows accordingly. Now the countries with the fewest requests appeared at the top and those with the most at the bottom, which made the heatmap cleaner and more intuitive to read.**

In [ ]:
selected_countries = ['Australia', 'Brazil', 'Canada', 'France', 'Germany', 
                      'India', 'Japan', 'South Korea', 'United Kingdom', 'United States']

df_selected = df[df['Country'].isin(selected_countries)]

country_totals = df_selected.groupby('Country')['Total Requests'].sum().reset_index()

country_totals = country_totals.sort_values(by='Total Requests', ascending=True)

sorted_countries = country_totals['Country'].tolist()

heatmap_data = df_selected.pivot_table(
    index='Country',
    columns='Year',
    values='Total Requests',
    aggfunc='sum'
)

heatmap_data = heatmap_data.loc[sorted_countries]


fig_heatmap1 = px.imshow(
    heatmap_data,
    labels=dict(x="Year", y="Country", color="Total Requests"),
    color_continuous_scale='RdYlBu_r',
    title='Heatmap of Total Requests by Country and Year'
)

fig_heatmap1.update_layout(
    title_x=0.50,
    xaxis_title="Year",
    yaxis_title="Country"
)

fig_heatmap1.show()


In [ ]:
fig_bar.show()

In [ ]:
fig_line.show()

In [ ]:
fig_map.show()

In [ ]:
fig_map_y.show()

In [ ]:
fig_rt.show()

In [ ]:
fig_heatmap.show()

In [ ]:
fig_heatmap1.show()

**My driving question for this project was to explore “How have global government requests to Facebook evolved over time, and what patterns exist in terms of country activity and request types?”.I wanted to explore how the volume and type of government requests such as legal process, emergency, and preservation requests varied across countries and years, and what that might reveal about surveillance trends, policy shifts,or geopolitical events.**

**I’ve been closely following how often governments around the world request data from Facebook, and the differences between countries are pretty striking. In 2024, the USA made over 158.839k requests the highest globally and a 20% increase since 2020.India stood out even more to me.Its requests almost doubled in just four years,going from around 39,692k in 2020 to over 80,004k in 2024—a massive 97% jump.**

**Brazil and Germany also showed sharp increases. Brazil’s requests grew by over 23.81k, and Germany saw an 87% rise.In contrast, countries like Japan, South Korea, and Australia stayed relatively low, averaging fewer than 2,000 requests per year.**

**When it comes to the types of requests, legal process requests were clearly the most common.They made up about 60–65% of the total and rose from 156,267 in 2020 to 240,762 in 2024,a 56% increase.Preservation requests made up around a quarter of the total and grew by 27%. Emergency requests were fewer in number but still rose significantly, increasing by 61%.**

**The heatmaps and choropleth visuals made it easier to grasp the global picture, especially how much activity is centered in countries like the USA and India.**

**Overall, global government data requests to Facebook jumped by 66% in just five years.But what stands out most to me is how unevenly this activity is spread. The USA and India alone account for more than half of all requests.It’s clear that governments are relying more and more on digital platforms for investigations and surveillance, but not all are doing so at the same pace.**

**In the end, I find it both fascinating and a bit concerning. The rapid rise in requests reflects how deeply digital platforms are embedded in modern governance and law enforcement.But the big differences between countries really make me think about issues like transparency, accountability, and personal privacy. As someone who’s genuinely interested in digital rights, I believe the way we handle these growing trends won’t just impact policy it’ll play a big role in deciding how we balance safety and freedom in our increasingly online world.**

**Targeted Audiances**
*While working on this project, I goal was to make the insights meaningful for people who care deeply about digital privacy, transparency, and tech policy. In my point of view, this dataset matters not just to researchers or analysts, but also to journalists, policy makers, and everyday users who want to understand how their data is being handled by platforms like Facebook.I focused on making the visuals interactive, clear, and easy to explore so anyone, whether technical or not, can quickly grasp who’s requesting user data, how it’s changing over time, and what that might mean for global digital rights.*

**Sources**
*The data was sourced directly from Meta’s official transparency reports, which are publicly available and widely regarded as authoritative for understanding government requests to Facebook.* **URL:- https://transparency.meta.com/reports/government-data-requests/**

**Quality:-**

*The data comes from Meta’s official reports, ensuring high credibility.Covers all countries for five consecutive years (2020–2024), with breakdowns by request type.Each year’s dataset follows the same structure, making merging and comparison straightforward.*

**Structure:-**

*The dataset is organized in Tabular format with each row represents a country in a given year. And has columns like Country, Year, Total Requests, Preservation Requests, Legal Process Requests, Emergency Requests, and their respective percentages and outcomes.*

**Collection:-**

*Downloaded annual CSVs for each year, 2020–2024, from Meta’s official website*

**Wrangling:-**

*I cleaned the data by removing columns I didn’t need (like FISA and NSL), fixing number formats, and filling in any missing values.Then, I merged all the years into one dataset and added a column for the year.After that, I used Python (mainly pandas and plotly) to explore trends and create visualizations.*

**Analysis:-**

*Here, I make sure that all the colunms were in correct format for accurate
aggregation and visulaization. Then, I grouped data by year and country to see trends overtime and across regions.And created subsets for focused analysis(eg.top 10 countries by request volume).*

**Visulazations:-**

*I chose Bar, Line, Choropleth, Heatmap because they are simple, clear, and clear to understand.For interactivity I used Dropdowns, hover data, and animations to make the charts more explorable and engaging, helping the audience discover insights themselves.*

**Bar plot:-**
*The bar plots shows the total number of request each year from 2020-2024*

**Line plot:-**
*In line plot, I wanted showed how each request type(Preservation, Legal, Emergency) evolved over time.*

**Choropleth Map:-**
*Here, in these map i summed all the years to show total request by country where dark blues highlighting the highest volume regions and the light blues shows the lowest volume regions*

**Animated Map:-**
*In these map, I wanted to make it interactive by created a slider where user can slide to different year to see how request activity changed year-by-year.

**Dropdown Interactive Map:-**
*Here, again user can switch between request types to compare their global distributions.*

**Heatmap:-**
*In heatmap,at first I tried to represent all the countries but when i run the code the graph looks ugly.So, I selected top 10 known countries where viewers can compare request activity across years with esay to read color scale*

**Reflection:-**

**Limitations:** *One big limitation is that the data doesn’t show what happened after each request like whether Facebook approved it, rejected it, or pushed back.That missing piece makes it hard to understand the full impact. Also, some countries might not show up in the dataset not necessarily because they didn’t make requests, but maybe because the data wasn’t reported or disclosed. I wanted to dig deeper and try coding a tool to trace patterns or outcomes based on available metadata, but without more detailed information, it’s tough to build something truly insightful.*

*In the future, I would like to focus on finding more detailed data about the outcomes of government requests such as whether they were approved, denied, or challenged. Since the current dataset doesn’t include that information, I’d need to explore additional sources, like transparency reports from Facebook itself or third-party digital rights organizations that track request responses. Combining these sources could help build a more complete picture. I’m also curious about comparing how different platforms( (like Google, Apple, or Twitter)handle these requests, which could make the analysis more comprehensive and insightful.*

*Moving forward, I want to take my code to the next level not just to display data, but to let users really explore it. I’m planning to add interactive filters, charts that respond to clicks, and comparisons across countries or platforms.I also want to pull in external sources, like transparency reports or public APIs, so I can go beyond just counting requests and start understanding what happens after them. Right now, I’m limited by the dataset, but with smarter code and better data, I believe I can turn this into a tool that actually reveals patterns and tells a story, something that’s not just informative, but meaningful.*